Replace bools in PDFContextMenuItem with enum classes.

sammygill opened the pull request:
#### 6e2b89f53e56d4171becce375bf908ce725e9d30
<pre>
Replace bools in PDFContextMenuItem with enum classes.
<a href="https://bugs.webkit.org/show_bug.cgi?id=265815.">https://bugs.webkit.org/show_bug.cgi?id=265815.</a>
<a href="https://rdar.apple.com/problem/119150073">rdar://problem/119150073</a>.

Reviewed by Simon Fraser.

enabled, separator, and hasAction are all bools which can be represented
as an enum class.

* Source/WebKit/Shared/mac/PDFContextMenu.h:
* Source/WebKit/Shared/mac/PDFContextMenuItem.serialization.in:
* Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:
(WebKit::WebPageProxy::showPDFContextMenu):
* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.mm:
(WebKit::PDFPlugin::handleContextMenuEvent):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::createContextMenu const):

Canonical link: <a href="https://commits.webkit.org/271554@main">https://commits.webkit.org/271554@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d6a64392ee5b426eee42146377c8ab66dac58a08

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/28699 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/7342 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/30070 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/31319 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/26186 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/9507 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/4703 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/26261 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/28969 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/6109 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/24681 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/5278 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/5460 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/32657 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/26307 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/26128 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/31669 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/5391 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/3579 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/29447 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/7008 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6879 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/5858 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/5909 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.mm:
```diff
@@ -1987,7 +1987,9 @@ static bool getEventTypeFromWebEvent(const WebEvent& event, NSEventType& eventTy
-        PDFContextMenuItem menuItem { String([item title]), !![item isEnabled], !![item isSeparatorItem], static_cast<int>([item state]), !![item action], i };
+        PDFContextMenuItem menuItem { String([item title]), static_cast<ContextMenuItemEnablement>(!![item isEnabled]),
+            static_cast<ContextMenuItemIsSeparator>(!![item isSeparatorItem]), static_cast<int>([item state]),
+            static_cast<ContextMenuItemHasAction>(!![item action]), i };
```
Isn't this UB? We didn't map the enumeration cases to literal values at the enum declaration.

```suggestion
        PDFContextMenuItem menuItem { String([item title]), [item isEnabled] ? ContextMenuItemEnablement::Enabled : ContextMenuItemEnablement::Disabled, [item isSeparatorItem] ? ContextMenuItemIsSeparator::Yes : ContextMenuItemIsSeparator::No, static_cast<int>([item state]), [item action] ? ContextMenuItemHasAction::Yes : ContextMenuItemHasAction::No, i };
```

sammygill replied:
I thought it was ok given the fixed underlying type but went ahead and added the values to the enum

cdumez replied:
I don't think this was UB? That said, Abrar's proposal looks nicer IMO so we may as well adopt it.

sammygill replied:
Changed and removed the values from the enum definitions

cdumez commented on Source/WebKit/Shared/mac/PDFContextMenu.h:
```diff
@@ -28,13 +28,17 @@
 #if ENABLE(PDF_PLUGIN) && PLATFORM(MAC)
 
 namespace WebKit {
+enum class ContextMenuItemEnablement : bool { Disabled = false, Enabled = true };
```
`= false` && `= true` are not needed.

cdumez commented on Source/WebKit/Shared/mac/PDFContextMenu.h:
```diff
@@ -28,13 +28,17 @@
 namespace WebKit {
+enum class ContextMenuItemEnablement : bool { Disabled = false, Enabled = true };
+
+enum class ContextMenuItemIsSeparator : bool { No = false, Yes = true };
```
ditto.

cdumez commented on Source/WebKit/Shared/mac/PDFContextMenu.h:
```diff
@@ -28,13 +28,17 @@
+
+enum class ContextMenuItemIsSeparator : bool { No = false, Yes = true };
     
+enum class ContextMenuItemHasAction : bool { No = false, Yes = true };
```
ditto.

smfr commented on Source/WebKit/Shared/mac/PDFContextMenu.h:
```diff
@@ -28,13 +28,17 @@
+    ContextMenuItemIsSeparator separator;
     int state;
-    bool hasAction;
+    ContextMenuItemHasAction hasAction;
```
Please group all the enums together for better padding.

sammygill replied:
Done. Also added replaced a `static_cast` with `enumToUnderlyingType()`

sammygill committed:
Replace bools in PDFContextMenuItem with enum classes.
https://bugs.webkit.org/show_bug.cgi?id=265815.
rdar://problem/119150073.

Reviewed by Simon Fraser.

enabled, separator, and hasAction are all bools which can be represented
as an enum class.

* Source/WebKit/Shared/mac/PDFContextMenu.h:
* Source/WebKit/Shared/mac/PDFContextMenuItem.serialization.in:
* Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:
(WebKit::WebPageProxy::showPDFContextMenu):
* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.mm:
(WebKit::PDFPlugin::handleContextMenuEvent):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::createContextMenu const):

Canonical link: https://commits.webkit.org/271554@main
